### ISSUE TITLE
samtools stat -p -t 12_3reads.bed fails on i686

### ISSUE THREAD

**Issue by zzaba** [samtools/samtools#1478]
Hi!
I was packaging `samtools` for `voidlinux` and noticed that if compiled for i686 architecture one of the tests fails, because
`/builddir/samtools-1.13/samtools stats /builddir/samtools-1.13/test/stat/12_overlaps.bam -p -t /builddir/samtools-1.13/test/stat/12_3reads.bed | tail -n+4`
produces file which differs from `12.3reads.nooverlap.expected` on line 21.
`SN	error rate:	3.928572e-02	# mismatches / bases mapped (cigar) | SN	error rate:	3.928571e-02	# mismatches / bases mapped (cigar)`
So basically error rate is 0.000001 less than expected. All tests pass for x86_64, aarch64, armv7l, x86_64-musl, armv6l-musl and aarch64-musl though.

**Comment by valeriuo** [comment]
That number is most likely not representable in 32 bit floats. The solution would probably be to trim the error rate precision down. Does this test failure affect your packaging effort?

**Comment by zzaba** [comment]
I was told that the failing test may be harmless but to at least report it upstream.
There's similar issue with htslib, one test failing because `tabix/vcf_file.vcf` differs from new on last line:
`4	3258448	.	TACACACAC	T	.	PASS	AN=4;AC=2	GT:GQ:DP	0/1:325:31	0/1:325:31 | 4	3258448	.	TACACACAC	T	nan	PASS	AN=4;AC=2	GT:GQ:DP	0/1:325:31	0/1:325:31`
nan instead of dot. bcftools fails 477 test though. So is samtools not meant to be used on 32 bit systems?

**Comment by valeriuo** [comment]
> So is samtools not meant to be used on 32 bit systems?

There was an effort to make HTSlib run on [32 bit platforms](https://github.com/samtools/htslib/pull/1301), but not one specific for SAMtools, since there was no request for it. I remember seeing that failed error rate test before the latest release, but it was fairly inconsequential to do anything about it. A quick hack would have been to reduce the number of decimals in the error rate output. The best approach would probably be to compare the numbers within a relative error margin, as there will always be floating point numbers that are representable on 64 bits, but not on 32 bits. So, to all intents and purposes, SAMtools can be considered 32-bit able.

**Comment by zzaba** [comment]
I see. Thanks for the prompt and informative replies!

**Comment by jmarshall** [comment]
The NaN vs dot thing is really a question of 387 (the default on 32-bit Intel) vs SSE (the default on 64-bit Intel) arithmetic. See samtools/hts-specs#145.
